File: problems/sorting/insertion_sort.java
package sorting;

import java.util.Arrays;

// for ascending order ------------
class Solution {
    public boolean visualize = false;

    public int[] InsertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }
            int key = arr[i];
            int j = i - 1;

            while ((j >= 0) && (arr[j] > key)) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
        return arr;
    }
}

// for descending order ----------------
class Solution2 {
    public boolean visualize = false;

    public int[] InsertionSortDes(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }
            int key = arr[i];
            int j = i - 1;

            while ((j >= 0) && (arr[j] < key)) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
        return arr;
    }
}

// for both ascending & descending -------------
class Solution3 {
    public boolean visualize = false;

    public int[] InsertionSort(int[] arr, boolean reverse) {
        for (int i = 1; i < arr.length; i++) {
            if (visualize) {
                System.out.println(Arrays.toString(arr));
            }
            int key = arr[i];
            int j = i - 1;

            while ((j >= 0) && (reverse ? (arr[j] < key) : (arr[j] > key))) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
        return arr;
    }
}

File: problems/sorting/insertion_sort.py
from typing import List

# for ascending order ---------------


class Solution:
    def InsertionSort(self, nums: List[int], visualize: bool = False) -> List[int]:
        for i in range(1, len(nums)):
            if visualize:
                print(nums)
            
            key = nums[i]
            j = i - 1
            
            while j >= 0 and nums[j] > key:
                nums[j + 1] = nums[j]
                j -= 1
                
            nums[j + 1] = key
        return nums



# for descending order -----------------

class Solution:
    def InsertionSortDes(self, nums: List[int], visualize: bool = False) -> List[int]:
        for i in range(1, len(nums)):
            if visualize:
                print(nums)
            
            key = nums[i]
            j = i - 1
            
            while j >= 0 and nums[j] < key:
                nums[j + 1] = nums[j]
                j -= 1
            
            nums[j + 1] = key
        return nums
    

# for both ascending & descending order --------------
class Solution:
    def InsertionSort(self, nums: List[int], reverse: bool = False, visualize: bool = False) -> List[int]:
        for i in range(1, len(nums)):
            if visualize:
                print(nums)
            
            key = nums[i]
            j = i - 1
            
            while j >= 0 and ((nums[j] < key) if reverse else (nums[j] > key)):
                nums[j + 1] = nums[j]
                j -= 1
            
            nums[j + 1] = key
        return nums

# obj = Solution()
# print(obj.InsertionSort([3, 2, 1, 2, 8, 9]))
